h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -583,6 +581,59 @@ public static String formatDateTime(java.util.Date date, String format,
         }
     }
 
+    @SuppressWarnings("serial")
+    private static final Map<String, String> DATE_FORMAT_REGEXPS = new HashMap<String, String>() {
+        {
+            put("^\\d{8}[^\\d]*", "yyyyMMdd");
+            put("^\\d{1,2}\\.\\d{1,2}\\.\\d{4}[^\\d]*", "dd.MM.yyyy");
+            put("^\\d{4}\\.\\d{1,2}\\.\\d{1,2}[^\\d]*", "yyyy.MM.dd");
+            put("^\\d{4}-\\d{1,2}-\\d{1,2}[^\\d]*", "yyyy-MM-dd");
+            put("^\\d{1,2}-\\d{1,2}-\\d{4}[^\\d]*", "dd-MM-yyyy");
+            put("^\\d{1,2}/\\d{1,2}/\\d{4}[^\\d]*", "dd/MM/yyyy");
+            put("^\\d{4}/\\d{1,2}/\\d{1,2}[^\\d]*", "yyyy/MM/dd");
+            put("^\\d{1,2}/\\d{1,2}/\\d{4}[^\\d]*", "dd/MM/yyyy");
+            put("^\\d{1,2}-[^\\d]{3}-\\d{4}[^\\d]*", "dd-MMM-yyyy");
+            put("^\\d{4}-[^\\d]{3}-\\d{1,2}[^\\d]*", "yyyy-MMM-dd");
+            put("^.{2}\\s.{3}\\s\\d{1,2}\\s\\d{1,2}\\:\\d{1,2}\\:\\d{1,2}\\s.{3}\\s\\d{4}$",
+                    "EEE MMM dd hh:mm:ss z yyyy");
+        }
+    };
+
+    /**
+     * Determine SimpleDateFormat pattern matching with the given date string. Returns null if format is unknown. You can
+     * simply extend DateUtil with more formats if needed.
+     *
+     * @param dateString
+     *          The date string to determine the SimpleDateFormat pattern for.
+     * @return The matching SimpleDateFormat pattern, or null if format is unknown.
+     */
+    private static String determineDateFormat(final String dateString) {
+        for (String regexp : DATE_FORMAT_REGEXPS.keySet()) {
+            if (dateString.toLowerCase().matches(regexp)) {
+                return DATE_FORMAT_REGEXPS.get(regexp);
+            }
+        }
+        return null; // Unknown format.
+    }
```

Comment:
this is potentially very expensive, repeatedly compiling a regular expression.
I suggest you compile them once in a static {} block and store the Pattern objects in the Map